### PR TITLE
Throw a syntax error for a constructor with type parameters

### LINKED ISSUE #12064: No error thrown for constructors with type parameters
## Bug Report

- [x] I would like to work on a fix!

**Current behavior**

[REPL Link](https://babeljs.io/repl#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAEDC0G8BQ1rAPYDsIBcBOArsDungDwAqAfABQBm66AXNBQJSIC-SnQA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=typescript&prettier=false&targets=&version=7.11.6&externalPlugins=)

No error thrown.

**Input Code**

```ts
class C {
  constructor<T>(foo: T) {}
}
```

**Expected behavior**

It throws Syntax Error `Type parameters cannot appear on a constructor declaration.` like TypeScript [TS Playground Link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDC0G8BQ1rAPYDsIBcBOArsDungDwAqAfABQBm66AXNBQJSIC+SnQA)

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
{
  "presets": ["@babel/preset-typescript"]
}
```

**Environment**
I don't think environment information is needed because this issue is simple, but I will provide it if it is needed for you guys.

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by sosukesuzuki**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12064
| Patch: Bug Fix?          | Maybe yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12065"></a>

### CODE REVIEW

File: scripts/parser-tests/typescript/allowlist.txt [hunk 1]
```diff
@@ -420,3 +420,4 @@ withStatement.ts
 withStatementErrors.ts
 withStatementInternalComments.ts
 withStatementNestedScope.ts
+parserConstructorDeclaration12.ts
```
Comment (sosukesuzuki) [inline scripts/parser-tests/typescript/allowlist.txt:423]: https://github.com/microsoft/TypeScript/blob/master/tests/cases/compiler/parserConstructorDeclaration12.ts
```ts
class C {
  constructor<>() { }
  constructor<> () { }
  constructor <>() { }
  constructor <> () { }
  constructor< >() { }
  constructor< > () { }
  constructor < >() { }
  constructor < > () { }
}
````

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -2291,6 +2293,9 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       allowsDirectSuper: boolean,
     ): void {
       const typeParameters = this.tsTryParseTypeParameters();
+      if (typeParameters && isConstructor) {
+        this.raise(method.start, TSErrors.ConstructorHasTypeParameters);
```
Comment (kaicataldo) [inline packages/babel-parser/src/plugins/typescript/index.js:2297]: It looks like TS reports this error on the [generic declaration](https://www.typescriptlang.org/play?#code/MYGwhgzhAEDC0G8BQ1rAPYDsIBcBOArsDungDwAqAfABQBm66AXNBQJSIC+SnQA). Should we try to match the error location?

Comment (sosukesuzuki) [reply packages/babel-parser/src/plugins/typescript/index.js]: Thank you. I've fixed:smile: 952ef30

**Review by kaicataldo — APPROVED** [review]
LGTM, thanks!

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

File: scripts/parser-tests/typescript/allowlist.txt (hunk 1 above)
Comment (nicolo-ribaudo) [reply scripts/parser-tests/typescript/allowlist.txt:423]: Can you update this line: https://github.com/babel/babel/blob/bbe0cf09fc60c130fc63a7176e563781582612b8/Makefile#L3
to update the TS tests?
And then run `make test-typescript-update-allowlist`

Comment (sosukesuzuki) [reply scripts/parser-tests/typescript/allowlist.txt]: I've fixed at c7fa9bd! But I'm not familiar with how babel-parser tests work, so I'm not sure if it's correct.

**Review by existentialism — APPROVED** [review]

**Comment by JLHwung** [comment]
The CI error is not related. I will address it in a separate PR.

File: scripts/parser-tests/typescript/allowlist.txt (hunk 1 above)
Comment (nicolo-ribaudo) [reply scripts/parser-tests/typescript/allowlist.txt]: It's correct! Btw, `allowlist` contain tests that we don't pass and not tests that we are correctly handling.

**Pull request merged**